ports/raspberrypi/common-hal/rp2pio/StateMachine.c

```diff
@@ -546,15 +546,23 @@ static bool _transfer(rp2pio_statemachine_obj_t *self,
         size_t tx_remaining = out_len;
 
         while (rx_remaining || tx_remaining) {
-            if (tx_remaining && !pio_sm_is_tx_fifo_full(self->pio, self->state_machine)) {
-                *tx_destination = *data_out;
-                data_out++;
-                --tx_remaining;
-            }
-            if (rx_remaining && !pio_sm_is_rx_fifo_empty(self->pio, self->state_machine)) {
-                *data_in = (uint8_t) *rx_source;
-                data_in++;
-                --rx_remaining;
+            for (int i=0; i<32; i++) {
+                bool did_transfer = false;
+                if (tx_remaining && !pio_sm_is_tx_fifo_full(self->pio, self->state_machine)) {
+                    *tx_destination = *data_out;
+                    data_out++;
+                    --tx_remaining;
+                    did_transfer = true;
+                }
+                if (rx_remaining && !pio_sm_is_rx_fifo_empty(self->pio, self->state_machine)) {
+                    *data_in = (uint8_t) *rx_source;
+                    data_in++;
+                    --rx_remaining;
+                    did_transfer = true;
+                }
+                if (!did_transfer) {
```

Comment:
I think it'd be clearer if you replaced the original if statements with while loops. That way you'd load up the 4 or 8 entry FIFO and then move on. 32 seems arbitrary and I'm skeptical you'd ever hit that many due to the FIFO size.

Reply:
I chose 32 because it was the DMA threshold.  I didn't check the FIFO size at all!  I'll open an issue suggesting the clean-up, but try to get to it next time I'm working in this code.